Mapping to an array of Object throws a NullPointerException

Proof of Concept code at the bottom of the post.

When a custom PropertyMap has been defined for a given Source and Destination, and that Destination has a setter that accepts an array of Objects (`Object[]`), a NullPointerException is thrown.

The NPE is thrown in `org.modelmapper.internal.util.Primitive` when processing the map. `Primitives.primitiveFor` is called once, to process the array, and then a second time, with the type parameter of the array. However, calls to `Primitives.primtivesFor` with a Object return null explicitly. The null value is then used as the type in an attempt to construct a new Array instance.

**affected version:** 0.7.8
**stack trace**

    Caused by: java.lang.NullPointerException
    	at java.lang.reflect.Array.newInstance(Array.java:313)
    	at org.modelmapper.internal.util.Primitives.primitiveFor(Primitives.java:144)
    	at org.modelmapper.internal.ExplicitMappingVisitor.methodFor(ExplicitMappingVisitor.java:332)
    	at org.modelmapper.internal.ExplicitMappingVisitor.access$500(ExplicitMappingVisitor.java:35)
    	at org.modelmapper.internal.ExplicitMappingVisitor$MappingCapturingVisitor.visitEnd(ExplicitMappingVisit    or.java:200)
    	at org.modelmapper.internal.asm.ClassReader.b(Unknown Source)
    	at org.modelmapper.internal.asm.ClassReader.accept(Unknown Source)
    	at org.modelmapper.internal.asm.ClassReader.accept(Unknown Source)
    	at org.modelmapper.internal.ExplicitMappingBuilder.visitPropertyMap(ExplicitMappingBuilder.java:222)
    	at org.modelmapper.PropertyMap.configure(PropertyMap.java:380)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:88)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
    	at java.lang.reflect.Method.invoke(Method.java:613)
    	at org.modelmapper.internal.ExplicitMappingBuilder.build(ExplicitMappingBuilder.java:195)

Proof of concept

    package modelmapper.test;
    
    import java.util.ArrayList;
    import java.util.List;
    
    import org.modelmapper.ModelMapper;
    import org.modelmapper.PropertyMap;
    
    public class ObjectArrayTest {
    
    	public static class CustomObject {
    		private String value;
    
    		public String getValue() {
    			return value;
    		}
    
    		public void setValue(String value) {
    			this.value = value;
    		}
    	}
    
    	public static class Source {
    		private List<CustomObject> objects;
    
    		public List<CustomObject> getObjects() {
    			return objects;
    		}
    
    		public void setObjects(List<CustomObject> objects) {
    			this.objects = objects;
    		}
    	}
    	
    	public static class Destination {
    		private CustomObject[] objects;
    
    		public CustomObject[] getObjects() {
    			return objects;
    		}
    
    		public void setObjects(CustomObject[] objects) {
    			this.objects = objects;
    		}
    	}
    	
    	public static void main (String[] args) {
    		ModelMapper mapper = new ModelMapper();
    		
    		Source source = new Source();
    		CustomObject obj = new CustomObject();
    		obj.setValue("value");
    		
    		List<CustomObject> objs = new ArrayList<CustomObject>();
    		objs.add(obj);
    		objs.add(obj);
    		objs.add(obj);
    		source.setObjects(objs);
    		
    		mapper.addMappings(new PropertyMap<Source, Destination>() {
    			@Override
    			protected void configure() {
    				map(source.getObjects()).setObjects(null);
    				
    			}
    		});
    		
    		Destination result = mapper.map(source, Destination.class);
    		mapper.validate();
    		
    		result.hashCode();
    	}
    
    }

--- a/core/src/main/java/org/modelmapper/internal/ExplicitMappingVisitor.java
+++ b/core/src/main/java/org/modelmapper/internal/ExplicitMappingVisitor.java
@@ -1,5 +1,6 @@
 package org.modelmapper.internal;
 
+import java.lang.reflect.Array;
 import java.lang.reflect.Field;
 import java.lang.reflect.Member;
 import java.lang.reflect.Method;
@@ -329,10 +330,39 @@ public class ExplicitMappingVisitor extends ClassVisitor {
     Type[] argumentTypes = methodType.getArgumentTypes();
     Class<?>[] paramTypes = new Class<?>[argumentTypes.length];
     for (int i = 0; i < paramTypes.length; i++) {
-      Class<?> paramType = Primitives.primitiveFor(argumentTypes[i]);
-      paramTypes[i] = paramType == null ? classFor(argumentTypes[i].getClassName()) : paramType;
+      paramTypes[i] = classFor(argumentTypes[i]);
     }
     return Members.methodFor(type, mn.name, paramTypes);
+  }
+
+  /**
+   * Returns a origin class for the ASM {@code type}, else {@code null}.
+   */
+  public Class<?> classFor(Type type) {
+    switch (type.getSort()) {
+      case Type.BOOLEAN:
+        return Boolean.TYPE;
+      case Type.CHAR:
+        return Character.TYPE;
+      case Type.BYTE:
+        return Byte.TYPE;
+      case Type.SHORT:
+        return Short.TYPE;
+      case Type.INT:
+        return Integer.TYPE;
+      case Type.LONG:
+        return Long.TYPE;
+      case Type.FLOAT:
+        return Float.TYPE;
+      case Type.DOUBLE:
+        return Double.TYPE;
+      case Type.ARRAY:
+        return Array.newInstance(classFor(type.getElementType()), new int[type.getDimensions()])
+            .getClass();
+      case Type.OBJECT:
+      default:
+        return classFor(type.getClassName());
+    }
   }
 
   private Class<?> classFor(String className) {

--- a/core/src/main/java/org/modelmapper/internal/util/Primitives.java
+++ b/core/src/main/java/org/modelmapper/internal/util/Primitives.java
@@ -15,13 +15,10 @@
  */
 package org.modelmapper.internal.util;
 
-import java.lang.reflect.Array;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
-
-import org.objectweb.asm.Type;
 
 /**
  * Utility methods for working with primitives.
@@ -117,36 +114,6 @@ public final class Primitives {
    */
   public static Class<?> primitiveFor(Class<?> wrapper) {
     return wrapperToPrimitive.get(wrapper);
-  }
-
-  /**
-   * Returns a primitive type for the ASM {@code type}, else {@code null}.
-   */
-  public static Class<?> primitiveFor(Type type) {
-    switch (type.getSort()) {
-      case Type.BOOLEAN:
-        return Boolean.TYPE;
-      case Type.CHAR:
-        return Character.TYPE;
-      case Type.BYTE:
-        return Byte.TYPE;
-      case Type.SHORT:
-        return Short.TYPE;
-      case Type.INT:
-        return Integer.TYPE;
-      case Type.LONG:
-        return Long.TYPE;
-      case Type.FLOAT:
-        return Float.TYPE;
-      case Type.DOUBLE:
-        return Double.TYPE;
-      case Type.ARRAY:
-        return Array.newInstance(primitiveFor(type.getElementType()), new int[type.getDimensions()])
-            .getClass();
-      case Type.OBJECT:
-      default:
-        return null;
-    }
   }
 
   /**

--- a/core/src/test/java/org/modelmapper/bugs/GH220.java
+++ b/core/src/test/java/org/modelmapper/bugs/GH220.java
@@ -1,0 +1,79 @@
+package org.modelmapper.bugs;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.modelmapper.AbstractTest;
+import org.modelmapper.PropertyMap;
+import org.testng.annotations.Test;
+
+@Test
+public class GH220 extends AbstractTest {
+  static class CustomObject {
+    private String value;
+
+    public String getValue() {
+      return value;
+    }
+
+    public void setValue(String value) {
+      this.value = value;
+    }
+  }
+
+  static class Source {
+    private List<CustomObject> objects;
+
+    public List<CustomObject> getObjects() {
+      return objects;
+    }
+
+    public void setObjects(List<CustomObject> objects) {
+      this.objects = objects;
+    }
+  }
+
+  static class Destination {
+    private CustomObject[] objects;
+
+    public CustomObject[] getObjects() {
+      return objects;
+    }
+
+    public void setObjects(CustomObject[] objects) {
+      this.objects = objects;
+    }
+  }
+
+  public void shouldMapNonPrimitiveArrayWithPropertyMap() throws Exception {
+    Source source = new Source();
+    CustomObject obj = new CustomObject();
+    obj.setValue("value");
+
+    List<CustomObject> objs = new ArrayList<CustomObject>();
+    objs.add(obj);
+    objs.add(obj);
+    objs.add(obj);
+    source.setObjects(objs);
+
+    modelMapper.addMappings(new PropertyMap<Source, Destination>() {
+      @Override
+      protected void configure() {
+        map(source.getObjects()).setObjects(null);
+
+      }
+    });
+    modelMapper.validate();
+
+    Destination result = modelMapper.map(source, Destination.class);
+
+    assertNotNull(result.objects);
+    assertEquals(result.objects.length, 3);
+    assertEquals(result.objects[0], obj);
+    assertEquals(result.objects[1], obj);
+    assertEquals(result.objects[2], obj);
+  }
+}

